### PR TITLE
fix(bridge): make native TransformerBridge state_dict()/load_state_dict() true inverses

### LINKED ISSUE #1587: [Bug Report] Native bridge state_dict()/load_state_dict() are not inverses
### **Base on `dev-4.x`**

**Describe the bug**

On a native `TransformerBridge`, `state_dict()` emits TL-renamed keys (`embed.weight`, `blocks.0.attn.q.weight`, …), but `load_state_dict()` only matches raw native names, so a `state_dict()` -> `load_state_dict()` round-trip silently loads nothing and `strict=True` is silently downgraded to `strict=False`. You can reload a checkpoint into a random-init model, get no error, and analyse garbage. This blocks the checkpoint save/reload workflow that `HookedTransformer` deprecation (removal at 4.0) requires.

**Code example**

```python
from transformer_lens.model_bridge import TransformerBridge

bridge = TransformerBridge.boot_transformers("gpt2")
import torch
with torch.no_grad():
    for p in bridge.parameters():
        p.zero_()
sd = bridge.state_dict()  # TL-renamed keys, e.g. "embed.weight"
bridge.load_state_dict(sd, strict=True)  # no-op; strict silently downgraded, no error
```

**System Info**

Installed from source; any OS; Python 3.10+. (Reproduces on the current `dev` / bridge.)

**Expected behaviour & fix pointer**

Round-trip must overwrite params, and `strict=True` must raise on genuine mismatch — mirroring `HookedTransformer`'s trivially-inverse `state_dict`/`load_state_dict`. In `transformer_lens/model_bridge/transformer_bridge.py`, give `load_state_dict` (`:3478-3510`) an inverse of `convert_hf_key_to_tl_key` (forward at `architecture_adapter.py:880`) before the clean↔actual mapping, and remove the silent strict downgrade at `:3507`. The forward direction is the `state_dict` override at `transformer_bridge.py:3424` (TL-key emission at `:3470`). Compatibility constraint: `transformer_lens/utilities/tracr.py` `make_tracr_transformer_bridge_state_dict` (def `:99`) emits raw native keys (`:117-122`; documented at `:106-108`) and relies on raw-name matching, so the inverse-key fix must keep that path loading.

Acceptance:
- [ ] `load_state_dict(state_dict())` overwrites params (mutate, round-trip, assert equal) — no no-op
- [ ] `strict=True` raises on mismatched state dict
- [ ] Existing tracr path (`utilities/tracr.py`, raw-native names) still loads
- [ ] New round-trip test under `tests/unit/model_bridge/`
- [ ] `make unit-test` and `uv run mypy .` pass

### Checklist
- [x] I have checked that there is no similar issue in the repo (**required**)

### PR DESCRIPTION
**Pull request by LightWork666**
Fixes #1587.

## The bug

On a native `TransformerBridge`, `state_dict()` returns TL-renamed keys (`embed.weight`, `blocks.0.attn.q.weight`, ...), but `load_state_dict()` only knew how to match raw native parameter names. So `bridge.load_state_dict(bridge.state_dict())` silently did nothing — no error, no warning, params just stayed whatever they were before. Worse, `strict=True` was silently downgraded to `strict=False` whenever the key counts didn't line up, so there was no way to even notice the round trip had failed.

## The fix

`load_state_dict` now builds the inverse of the TL-key renaming that `state_dict()` applies (`_tl_key_to_actual_keys`), so it can map TL-format keys back to the underlying native parameter paths before handing them to the wrapped model's own `load_state_dict`. One wrinkle: some bridge components expose the same underlying parameter through more than one attribute path — e.g. GPT-2's `q`/`k`/`v` are views into the wrapped module's combined `c_attn` weight, reachable both through a block-level shortcut and through the nested `_original_component` chain. Writing to only one of those paths leaves the model's actual forward pass untouched even though `state_dict()` looks fine, so the mapping keeps every alias for a given TL key, not just the first one found.

The silent `strict=True → False` downgrade is gone. Missing/unexpected keys are now computed properly (scoped to what the TL state dict actually needs) and raise a real `RuntimeError` under `strict=True`, matching how `HookedTransformer` already behaves.

The raw-key loading path used by `tracr` (`make_tracr_transformer_bridge_state_dict`) still works — raw native keys are matched directly before falling through to the TL-key path.

## Testing

- New tests in `tests/unit/model_bridge/test_state_dict_round_trip.py`: round-trip actually overwrites params (not a no-op), `strict=True` raises on both missing and unexpected keys, `strict=False` doesn't raise on a partial dict, the tracr raw-key path still loads, and a real GPT-2 case checking forward-pass logits match after a zero-and-reload cycle (this last one is what caught the aliasing issue above — a naive key-rename fix passes the round-trip-key-equality check but still produces different logits, because it never touches the aliased storage).
- `uv run mypy .` — clean.
- Ran the full `tests/unit/model_bridge/` suite (207 files) in isolated batches; everything passes except a handful of pre-existing `generate()`/KV-cache crashes already tracked as an upstream PyTorch/HF bug on Apple Silicon in `tests/QUARANTINES.md` — confirmed via `git stash` that those reproduce identically on unmodified `dev-4.x`, unrelated to this change.

## Note on #1595

I noticed after finishing this that #1595 is already open for the same issue, taking a related but different approach (it detects aliasing by comparing live tensor identity rather than by TL-key-name collisions, which is arguably a more principled check). I'm submitting this anyway since it was already done and transparency seemed better than not mentioning it. Happy to have the maintainers pick whichever they prefer, or close this if #1595 is the better fix.

### CODE REVIEW

File: transformer_lens/model_bridge/transformer_bridge.py [hunk 1]
```diff
@@ -3494,26 +3524,47 @@ def load_state_dict(self, state_dict, strict=True, assign=False):
         """
         current_state_dict = self.original_model.state_dict()
         clean_to_actual = {}
-        actual_to_clean = {}
         for actual_key in current_state_dict.keys():
             if actual_key != "_original_component":
-                clean_key = actual_key.replace("._original_component", "")
-                clean_to_actual[clean_key] = actual_key
-                actual_to_clean[actual_key] = clean_key
+                clean_to_actual[actual_key.replace("._original_component", "")] = actual_key
```
Comment (jlarson4) [inline transformer_lens/model_bridge/transformer_bridge.py:3529]: A complete raw-HF-format gpt2 checkpoint raises under `strict=True` (337 missing) even though `strict=False` restores the forward pass exactly. The clean branch maps one alias while `required_actual_keys` demands the union (lines 3546-3549). Can the strict accounting treat a required key as satisfied when any alias of its tensor is written?

File: tests/unit/model_bridge/test_state_dict_round_trip.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            p.zero_()
+
+    bridge.load_state_dict(raw_sd, strict=True)
+
+    reloaded_raw = bridge.original_model.state_dict()
+    for key, value in raw_sd.items():
+        assert torch.equal(reloaded_raw[key], value), f"{key} did not round-trip"
+
+
+@pytest.mark.slow
```
Comment (jlarson4) [inline tests/unit/model_bridge/test_state_dict_round_trip.py:146]: This is the only test in any of the three competing PRs that exercises real HF-key conversion, but slow is deselected by make unit-test and the MPS job only runs for main-targeted PRs. Can you add an unmarked `boot_native` test that builds a multi-alias clean-key dict at `strict=True`?

**Review by jlarson4 — COMMENTED** [review]
Thanks for taking a look at this @LightWork666! My intention is to take this implementation as our primary solution, and keep a couple tests and other small elements from #1591 and #1595. I do have a couple change requests, please take a look when you have a moment. Your hard work on this is much appreciated

File: transformer_lens/model_bridge/transformer_bridge.py (hunk 1 above)
Comment (LightWork666) [reply transformer_lens/model_bridge/transformer_bridge.py:3529]: You're right, thanks for catching this. The bug was that `required_actual_keys` demanded every alias in a TL key's group be present, but the clean-key branch only ever writes one alias per key. Since aliases of the same TL key share the underlying storage, writing any one of them already updates what forward() reads for all of them, so treating the group as satisfied by any single alias is correct. Fixed in the latest commit: `missing_keys` now checks `any(k in mapped_state_dict for k in actual_keys)` per group instead of requiring the full union. Verified against your exact repro (gpt2, clean-key dict, strict=True): 0 missing, 0 unexpected now.

File: tests/unit/model_bridge/test_state_dict_round_trip.py (hunk 2 above)
Comment (LightWork666) [reply tests/unit/model_bridge/test_state_dict_round_trip.py:146]: Added `test_native_clean_key_dict_with_partial_aliases_does_not_raise_strict`. No HF download needed; it uses `boot_native`'s own `_original_component` wrapping, which turns out to alias internally too (not just gpt2's c_attn split), so it exercises the same bug without needing the slow tier. Also added a second slow test (`test_boot_transformers_clean_key_dict_does_not_raise_strict`) that's your exact repro on real gpt2, so there's a permanent regression test for the specific case you hit too.

**Comment by jlarson4** [comment]
@LightWork666 thanks for the updates! Looks good, merging now
